IdSSLOpenSSL.pas

```diff
@@ -3225,171 +3528,231 @@ procedure TIdSSLContext.DestroyContext;
+  begin
+    if not(sslvSSLv2 in SSLVersions) then
+    begin
       SSL_CTX_set_options(fContext, SSL_OP_NO_SSLv2);
     end
-    else if (fMethod = sslvSSLv23) then begin
+    else if (fMethod = sslvSSLv23) then
+    begin
       SSL_CTX_clear_options(fContext, SSL_OP_NO_SSLv2);
     end;
   end;
   // SSLv3 might also be disabled as well..
-  if IsOpenSSL_SSLv3_Available then begin
-    if not (sslvSSLv3 in SSLVersions) then begin
+  if IsOpenSSL_SSLv3_Available then
+  begin
+    if not(sslvSSLv3 in SSLVersions) then
+    begin
       SSL_CTX_set_options(fContext, SSL_OP_NO_SSLv3);
     end
-    else if (fMethod = sslvSSLv23) then begin
+    else if (fMethod = sslvSSLv23) then
+    begin
       SSL_CTX_clear_options(fContext, SSL_OP_NO_SSLv3);
     end;
   end;
   // may as well do the same for all of them...
-  if IsOpenSSL_TLSv1_0_Available then begin
-    if not (sslvTLSv1 in SSLVersions) then begin
+  if IsOpenSSL_TLSv1_0_Available then
+  begin
+    if not(sslvTLSv1 in SSLVersions) then
+    begin
       SSL_CTX_set_options(fContext, SSL_OP_NO_TLSv1);
     end
-    else if (fMethod = sslvSSLv23) then begin
+    else if (fMethod = sslvSSLv23) then
+    begin
       SSL_CTX_clear_options(fContext, SSL_OP_NO_TLSv1);
     end;
   end;
-{IMPORTANT!!!  Do not set SSL_CTX_set_options SSL_OP_NO_TLSv1_1 and
-SSL_OP_NO_TLSv1_2 if that functionality is not available.  OpenSSL 1.0 and
-earlier do not support those flags.  Those flags would only cause
-an invalid MAC when doing SSL.}
-  if IsOpenSSL_TLSv1_1_Available then begin
-    if not (sslvTLSv1_1 in SSLVersions) then begin
+  { IMPORTANT!!!  Do not set SSL_CTX_set_options SSL_OP_NO_TLSv1_1 and
+    SSL_OP_NO_TLSv1_2 if that functionality is not available.  OpenSSL 1.0 and
+    earlier do not support those flags.  Those flags would only cause
+    an invalid MAC when doing SSL. }
+  if IsOpenSSL_TLSv1_1_Available then
+  begin
+    if not(sslvTLSv1_1 in SSLVersions) then
+    begin
       SSL_CTX_set_options(fContext, SSL_OP_NO_TLSv1_1);
     end
-    else if (fMethod = sslvSSLv23) then begin
+    else if (fMethod = sslvSSLv23) then
+    begin
       SSL_CTX_clear_options(fContext, SSL_OP_NO_TLSv1_1);
     end;
   end;
-  if IsOpenSSL_TLSv1_2_Available then begin
-    if not (sslvTLSv1_2 in SSLVersions) then begin
+  if IsOpenSSL_TLSv1_2_Available then
+  begin
+    if not(sslvTLSv1_2 in SSLVersions) then
+    begin
       SSL_CTX_set_options(fContext, SSL_OP_NO_TLSv1_2);
     end
-    else if (fMethod = sslvSSLv23) then begin
+    else if (fMethod = sslvSSLv23) then
+    begin
       SSL_CTX_clear_options(fContext, SSL_OP_NO_TLSv1_2);
     end;
   end;
+  if IsOpenSSL_TLSv1_3_Available then begin
+    if not(sslvTLSv1_3 in SSLVersions) then begin
+      SSL_CTX_set_options(fContext, SSL_OP_NO_TLSv1_3);
+    end
+	else if (fMethod = sslvSSLv23) then begin
+      SSL_CTX_clear_options(fContext, SSL_OP_NO_TLSv1_3);
+    end;
+  end;
 
+  if sslvTLSv1_3 in SSLVersions then

```

Comment:
This set of `if`s feels backwards. Let's say `SSLVersions` is set to `[sslvTLSv1..sslvTLSv1_3]` (not unreasonable given that many people use `[sslvTLSv1..sslvTLSv1_2]`), then this series of `if`s would set the acceptable proto range to 1.3 only. But if the `if`s are swapped so that `sslvTLSv1` is tested first and `sslvTLSv1_3` last, then the proto range would be set to 1.0-1.3 instead, as the user intended.